components/firebase_admin_sdk/sources/new-doc-in-firestore-collection/new-doc-in-firestore-collection.js

```diff
@@ -0,0 +1,120 @@
+  name: "New Document in Firestore Collection",
+  description: "Emits an event when a structured query returns new documents",
+  version: "0.0.1",
+  dedupe: "unique",
+  props: {
+    ...common.props,
+    db: "$.service.db",
+    apiKey: {
+      propDefinition: [
+        common.props.firebase,
+        "apiKey",
+      ],
+    },
+    query: {
+      propDefinition: [
+        common.props.firebase,
+        "query",
+      ],
+    },
+  },
+  hooks: {
+    ...common.hooks,
+    async deploy() {
+      await this.firebase.initializeApp();
+      const {
+        idToken,
+        refreshToken,
+        expiresIn,
+      } = await this.firebase.getToken(
+        this.apiKey,
+      );
+      this.setTokenInfo(idToken, refreshToken, expiresIn);
+      await this.firebase.deleteApp();
+    },
+  },
+  methods: {
+    ...common.methods,
+    _getToken() {
+      return this.db.get("token");
+    },
+    _setToken(token) {
+      this.db.set("token", token);
+    },
+    _getRefreshToken() {
+      return this.db.get("refreshToken");
+    },
+    _setRefreshToken(refreshToken) {
+      this.db.set("refreshToken", refreshToken);
+    },
+    _getExpires() {
+      return this.db.get("expires");
+    },
+    _setExpires(expires) {
+      this.db.set("expires", expires);
+    },
+    getExpiresTime(expires, timestamp = null) {
+      if (timestamp) return (timestamp * 1000) + (expires * 1000);
+      else return Date.now() + expires * 1000;
+    },
+    setTokenInfo(token, refreshToken, expires, timestamp = null) {
+      this._setToken(token);
+      this._setRefreshToken(refreshToken);
+      this._setExpires(this.getExpiresTime(expires, timestamp));
+    },
+    generateMeta({ document }) {
+      const {
+        name,
+        createTime,
+      } = document;
+      const id = name.substring(name.lastIndexOf("/") + 1);
+      return {
+        id,
+        summary: name,
+        ts: Date.parse(createTime),
+      };
+    },
+  },
+  async run(event) {
+    // refresh the authorization token if it will expire before the next run
+    const {
+      timestamp,
+      interval_seconds: intervalSeconds,
+    } = event;
+    if (intervalSeconds) {
+      const intervalMs = 1000 * intervalSeconds;
+      const nextRun = (timestamp * 1000) + intervalMs;
+      const expires = this._getExpires("expires");
+      if (expires <= nextRun) {
+        const refreshToken = this._getRefreshToken("refreshToken");
+        const {
+          id_token: newToken,
+          refresh_token: newRefreshToken,
+          expires_in: expiresIn,
+        } = await this.firebase.refreshToken(this.apiKey, refreshToken);
+        this.setTokenInfo(newToken, newRefreshToken, expiresIn, timestamp);
+      }
+    }
+
+    const token = this._getToken();
+    const { projectId } = this.firebase.$auth;
+    const parent = `projects/${projectId}/databases/(default)/documents`;
+    const structuredQuery = JSON.parse(this.query);
```

Comment:
Any chance we could encapsulate this in the `.app.js` module?
Especially since future event sources might also be calling the REST API directly.

I would say:
1. Let's keep the refresh token logic in a "private" method in the `.app.js` module.
2. Make the `firebase.runQuery` method accept only the `structuredQuery` argument (and possibly a `collection` argument, like Zapier does)
3. The `firebase.runQuery` would then refresh the token by calling the method from step 1, and then it will make the REST API call.

The reason I'm pushing for this is that it will keep all the Firebase client logic inside the `.app.js` module, and it will make it easily reusable for future event sources.

Thoughts?

Reply:
Moved token logic to the app file. I added a "global" variable called tokens to the app file to hold onto the tokens between executions and avoid passing the tokens back and forth between the components and the app file. Let me know if there's a better way I should be doing that.